security/advancedtls/advancedtls.go

```diff
@@ -0,0 +1,375 @@
+	RootCACerts *x509.CertPool
+	// If GetRootCAs is set and RootCACerts is nil, the server will invoke this function every time
+	// asked to check certificates sent from the client when a new connection is established.
+	// This is known as trust certificate reloading.
+	GetRootCAs func(rawConn net.Conn, rawCerts [][]byte) (*x509.CertPool, error)
+	// If the server want the client to send certificates to prove its identity.
+	MutualAuth bool
+}
+
+func (o *ClientOptions) config() (*tls.Config, error) {
+	if o.RootCACerts == nil && o.GetRootCAs == nil {
+		return nil, fmt.Errorf("either RootCACerts or GetRootCAs must be specified")
+	}
+	config := &tls.Config{
+		ServerName:           o.ServerNameOverride,
+		Certificates:         o.Certificates,
+		GetClientCertificate: o.GetClientCertificate,
+		RootCAs:              o.RootCACerts,
+	}
+	// We fall back to normal config settings if users don't need any of the advanced TLS features.
+	if o.GetClientCertificate == nil && o.GetRootCAs == nil && o.VerifyPeer == nil {
+		config.InsecureSkipVerify = false
+	} else {
+		config.InsecureSkipVerify = true
+	}
+	return config, nil
+}
+
+func (o *ServerOptions) config() (*tls.Config, error) {
+	if o.Certificates == nil && o.GetCertificate == nil {
+		return nil, fmt.Errorf("either Certificates or GetCertificate must be specified")
+	}
+	if !o.MutualAuth {
+		return &tls.Config{
+			ClientAuth:     tls.NoClientCert,
+			Certificates:   o.Certificates,
+			GetCertificate: o.GetCertificate,
+		}, nil
+	}
+	if o.GetRootCAs == nil && o.RootCACerts == nil {
+		return nil, fmt.Errorf("server needs trust certs if using mutual TLS")
+	}
+	// We fall back to normal config settings if users don't need to reload root certificates.
+	if o.RootCACerts != nil {
+		return &tls.Config{
+			ClientAuth:     tls.RequireAndVerifyClientCert,
+			Certificates:   o.Certificates,
+			GetCertificate: o.GetCertificate,
+			ClientCAs:      o.RootCACerts,
+		}, nil
+	}
+	return &tls.Config{
+		ClientAuth:     tls.RequireAnyClientCert,
+		Certificates:   o.Certificates,
+		GetCertificate: o.GetCertificate,
+	}, nil
+
+}
+
+// advancedTLSCreds is the credentials required for authenticating a connection using TLS.
+type advancedTLSCreds struct {
+	config     *tls.Config
+	verifyFunc CustomVerificationFunc
+	getRootCAs func(rawConn net.Conn, rawCerts [][]byte) (*x509.CertPool, error)
+	isClient   bool
+}
+
+func (c advancedTLSCreds) Info() credentials.ProtocolInfo {
+	return credentials.ProtocolInfo{
+		SecurityProtocol: "tls",
+		SecurityVersion:  "1.2",
+		ServerName:       c.config.ServerName,
+	}
+}
+
+func (c *advancedTLSCreds) ClientHandshake(ctx context.Context, authority string,
+	rawConn net.Conn) (_ net.Conn, _ credentials.AuthInfo, err error) {
+	// use local cfg to avoid clobbering ServerName if using multiple endpoints
+	cfg := cloneTLSConfig(c.config)
+	if cfg.ServerName == "" {
+		colonPos := strings.LastIndex(authority, ":")
```

Comment:
why you strip character after ":"? It is better to pass authority as it is to CustomVerificationFunc, so that application can decide? It is not gRPC transport credentials library's job to strip the trailing port.

Reply:
Yeah I agree. We could pass the whole authority to user for custom stripping.